// amvault-connect/dist/eip1193/AmvaultEIP1193Provider.js
// eip1193/AmvaultEIP1193Provider.ts
//
// EIP-1193 compatible provider that wraps AmVault's popup flows.
// Lets dApps use standard ethers.js / wagmi / viem without any
// AmVault-specific code:
//
//   import { AmvaultEIP1193Provider } from 'amvault-connect'
//   const provider = new AmvaultEIP1193Provider({ appName, amvaultUrl, chainId })
//   const ethers  = new BrowserProvider(provider)             // ethers v6
//   const [addr]  = await ethers.send('eth_requestAccounts', [])
import { openSignin, signMessage, sendTransaction } from '../popup/amvaultProvider';
import { makeStorageKeys } from '../provider/AuthProvider';
// Standard EIP-1193 error codes
const RPC_ERROR = {
    USER_REJECTED: 4001,
    UNAUTHORIZED: 4100,
    UNSUPPORTED_METHOD: 4200,
    DISCONNECTED: 4900,
    CHAIN_DISCONNECTED: 4901,
    INTERNAL: -32603,
    INVALID_PARAMS: -32602,
    METHOD_NOT_FOUND: -32601,
};
function rpcError(message, code, data) {
    const err = new Error(message);
    err.code = code;
    if (data !== undefined)
        err.data = data;
    return err;
}
// ─── Provider ──────────────────────────────────────────────────────────────
export class AmvaultEIP1193Provider {
    constructor(config) {
        var _a;
        this.isAmvault = true;
        this.isMetaMask = false; // don't masquerade as MetaMask
        this._listeners = {};
        this._config = config;
        this._storageKeys = makeStorageKeys((_a = config.storagePrefix) !== null && _a !== void 0 ? _a : 'amvault');
    }
    // ─── Session helpers ───────────────────────────────────────────────────
    _loadSession() {
        try {
            const raw = localStorage.getItem(this._storageKeys.session);
            if (!raw)
                return null;
            const s = JSON.parse(raw);
            if (Date.now() >= s.expiresAt) {
                localStorage.removeItem(this._storageKeys.session);
                return null;
            }
            return s;
        }
        catch {
            return null;
        }
    }
    _saveSession(session) {
        localStorage.setItem(this._storageKeys.session, JSON.stringify(session));
    }
    _clearSession() {
        localStorage.removeItem(this._storageKeys.session);
    }
    // ─── EIP-1193 request() ───────────────────────────────────────────────
    async request(args) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p;
        const { method, params } = args;
        const p = (params !== null && params !== void 0 ? params : []);
        if (this._config.debug) {
            console.log('[AmvaultEIP1193] request', method, p);
        }
        switch (method) {
            // ── Read-only / no popup needed ──────────────────────────────────
            case 'eth_chainId': {
                return '0x' + this._config.chainId.toString(16);
            }
            case 'net_version': {
                return String(this._config.chainId);
            }
            case 'eth_accounts': {
                const session = this._loadSession();
                return session ? [session.address.toLowerCase()] : [];
            }
            // ── Auth / connect ───────────────────────────────────────────────
            case 'eth_requestAccounts': {
                // Return existing session if still valid
                const existing = this._loadSession();
                if (existing)
                    return [existing.address.toLowerCase()];
                // Otherwise open sign-in popup
                const nonce = this._makeNonce();
                const origin = window.location.origin;
                const msg = this._buildMessage(nonce);
                let resp;
                try {
                    resp = await openSignin({
                        app: this._config.appName,
                        chainId: this._config.chainId,
                        origin,
                        nonce,
                        amvaultUrl: this._config.amvaultUrl,
                        debug: !!this._config.debug,
                        message: msg,
                    });
                }
                catch (e) {
                    if (/reject|cancel|blocked/i.test((_a = e === null || e === void 0 ? void 0 : e.message) !== null && _a !== void 0 ? _a : '')) {
                        throw rpcError('User rejected the request', RPC_ERROR.USER_REJECTED);
                    }
                    throw rpcError((_b = e === null || e === void 0 ? void 0 : e.message) !== null && _b !== void 0 ? _b : 'Sign-in failed', RPC_ERROR.INTERNAL);
                }
                if (!(resp === null || resp === void 0 ? void 0 : resp.ok)) {
                    throw rpcError((_c = resp === null || resp === void 0 ? void 0 : resp.error) !== null && _c !== void 0 ? _c : 'Sign-in rejected', RPC_ERROR.USER_REJECTED);
                }
                // Verify response integrity
                const { verifyMessage } = await import('ethers');
                const toVerify = typeof resp.message === 'string' && resp.message.trim()
                    ? resp.message : msg;
                const recovered = verifyMessage(toVerify, resp.signature).toLowerCase();
                if (recovered !== resp.address.toLowerCase()) {
                    throw rpcError('Signature verification failed', RPC_ERROR.INTERNAL);
                }
                if (resp.nonce !== nonce) {
                    throw rpcError('Nonce mismatch', RPC_ERROR.INTERNAL);
                }
                if (Number(resp.chainId) !== this._config.chainId) {
                    throw rpcError(`Wrong network: got ${resp.chainId}, expected ${this._config.chainId}`, RPC_ERROR.CHAIN_DISCONNECTED);
                }
                // Resolve AIN
                let ain = '';
                if (typeof resp.ain === 'string' && resp.ain.trim())
                    ain = resp.ain.trim();
                else if (typeof resp.amid === 'string' && resp.amid.trim())
                    ain = resp.amid.trim();
                if (!ain && ((_d = this._config.registry) === null || _d === void 0 ? void 0 : _d.getAin)) {
                    try {
                        const gotAin = await this._config.registry.getAin(resp.address);
                        if (gotAin)
                            ain = gotAin;
                    }
                    catch { }
                }
                if (!ain)
                    ain = `ain-${resp.address.slice(2, 8)}`;
                const ttl = (_e = this._config.sessionTtlMs) !== null && _e !== void 0 ? _e : 86400000;
                const now = Date.now();
                const session = {
                    ain,
                    address: resp.address,
                    issuedAt: now,
                    expiresAt: now + ttl,
                };
                this._saveSession(session);
                this._emit('accountsChanged', [resp.address.toLowerCase()]);
                this._emit('connect', { chainId: '0x' + this._config.chainId.toString(16) });
                return [resp.address.toLowerCase()];
            }
            // ── Message signing ──────────────────────────────────────────────
            case 'personal_sign': {
                // EIP-1193: params = [message, address]
                // (some libs send [address, message] — we detect and normalise)
                const session = this._loadSession();
                if (!session) {
                    throw rpcError('Not connected. Call eth_requestAccounts first.', RPC_ERROR.UNAUTHORIZED);
                }
                // Normalise param order: ethers sends [msg, addr], wagmi sends [addr, msg]
                let rawMsg = p[0];
                if (typeof p[0] === 'string' &&
                    p[0].startsWith('0x') &&
                    p[0].length === 42 &&
                    typeof p[1] === 'string') {
                    rawMsg = p[1];
                }
                // Decode hex-encoded message to UTF-8 string if needed
                const message = this._decodeMessage(rawMsg);
                try {
                    return await signMessage({ chainId: this._config.chainId, message }, {
                        app: this._config.appName,
                        amvaultUrl: this._config.amvaultUrl,
                        debug: !!this._config.debug,
                    });
                }
                catch (e) {
                    if (/reject|cancel/i.test((_f = e === null || e === void 0 ? void 0 : e.message) !== null && _f !== void 0 ? _f : '')) {
                        throw rpcError('User rejected the request', RPC_ERROR.USER_REJECTED);
                    }
                    throw rpcError((_g = e === null || e === void 0 ? void 0 : e.message) !== null && _g !== void 0 ? _g : 'Sign failed', RPC_ERROR.INTERNAL);
                }
            }
            case 'eth_sign': {
                // eth_sign: params = [address, message]
                const session = this._loadSession();
                if (!session) {
                    throw rpcError('Not connected. Call eth_requestAccounts first.', RPC_ERROR.UNAUTHORIZED);
                }
                const rawMsg = p[1];
                const message = this._decodeMessage(rawMsg);
                try {
                    return await signMessage({ chainId: this._config.chainId, message }, {
                        app: this._config.appName,
                        amvaultUrl: this._config.amvaultUrl,
                        debug: !!this._config.debug,
                    });
                }
                catch (e) {
                    if (/reject|cancel/i.test((_h = e === null || e === void 0 ? void 0 : e.message) !== null && _h !== void 0 ? _h : '')) {
                        throw rpcError('User rejected the request', RPC_ERROR.USER_REJECTED);
                    }
                    throw rpcError((_j = e === null || e === void 0 ? void 0 : e.message) !== null && _j !== void 0 ? _j : 'Sign failed', RPC_ERROR.INTERNAL);
                }
            }
            // ── Transactions ─────────────────────────────────────────────────
            case 'eth_sendTransaction': {
                const session = this._loadSession();
                if (!session) {
                    throw rpcError('Not connected. Call eth_requestAccounts first.', RPC_ERROR.UNAUTHORIZED);
                }
                const tx = p[0];
                if (!tx || typeof tx !== 'object') {
                    throw rpcError('Invalid transaction object', RPC_ERROR.INVALID_PARAMS);
                }
                // Convert hex gas values to numbers if needed
                const gas = tx.gas ? parseInt(tx.gas, 16) : undefined;
                const gasPrice = tx.gasPrice ? tx.gasPrice : undefined;
                try {
                    return await sendTransaction({
                        chainId: this._config.chainId,
                        to: tx.to,
                        data: tx.data,
                        value: (_k = tx.value) !== null && _k !== void 0 ? _k : 0,
                        gas,
                        // gasPrice passed through as-is; AmVault handles EIP-1559 fields natively
                    }, {
                        app: this._config.appName,
                        amvaultUrl: this._config.amvaultUrl,
                        debug: !!this._config.debug,
                    });
                }
                catch (e) {
                    if (/reject|cancel/i.test((_l = e === null || e === void 0 ? void 0 : e.message) !== null && _l !== void 0 ? _l : '')) {
                        throw rpcError('User rejected the request', RPC_ERROR.USER_REJECTED);
                    }
                    throw rpcError((_m = e === null || e === void 0 ? void 0 : e.message) !== null && _m !== void 0 ? _m : 'Transaction failed', RPC_ERROR.INTERNAL);
                }
            }
            // ── Chain / network ──────────────────────────────────────────────
            case 'wallet_switchEthereumChain': {
                // Alkebuleum is single-chain for now; accept only our own chainId
                const requested = p[0];
                const requestedId = parseInt((_o = requested === null || requested === void 0 ? void 0 : requested.chainId) !== null && _o !== void 0 ? _o : '0', 16);
                if (requestedId !== this._config.chainId) {
                    throw rpcError(`AmVault only supports chainId ${this._config.chainId}`, RPC_ERROR.CHAIN_DISCONNECTED);
                }
                return null; // EIP-1193: null = success
            }
            case 'wallet_addEthereumChain': {
                // If the chainId matches ours, silently accept; otherwise reject
                const chain = p[0];
                const requestedId = parseInt((_p = chain === null || chain === void 0 ? void 0 : chain.chainId) !== null && _p !== void 0 ? _p : '0', 16);
                if (requestedId !== this._config.chainId) {
                    throw rpcError(`AmVault only supports chainId ${this._config.chainId}`, RPC_ERROR.UNSUPPORTED_METHOD);
                }
                return null;
            }
            // ── Unsupported ──────────────────────────────────────────────────
            default: {
                throw rpcError(`AmVault does not support method: ${method}`, RPC_ERROR.UNSUPPORTED_METHOD);
            }
        }
    }
    // ─── EIP-1193 event emitter ────────────────────────────────────────────
    on(event, listener) {
        if (!this._listeners[event]) {
            this._listeners[event] = new Set();
        }
        this._listeners[event].add(listener);
        return this;
    }
    removeListener(event, listener) {
        var _a;
        (_a = this._listeners[event]) === null || _a === void 0 ? void 0 : _a.delete(listener);
        return this;
    }
    off(event, listener) {
        return this.removeListener(event, listener);
    }
    _emit(event, ...args) {
        var _a;
        (_a = this._listeners[event]) === null || _a === void 0 ? void 0 : _a.forEach((fn) => {
            try {
                fn(...args);
            }
            catch { }
        });
    }
    // ─── Disconnect (clears session) ───────────────────────────────────────
    disconnect() {
        this._clearSession();
        this._emit('accountsChanged', []);
        this._emit('disconnect', rpcError('Disconnected', RPC_ERROR.DISCONNECTED));
    }
    // ─── Private helpers ───────────────────────────────────────────────────
    _makeNonce() {
        const b = new Uint8Array(16);
        crypto.getRandomValues(b);
        return Array.from(b).map((x) => x.toString(16).padStart(2, '0')).join('');
    }
    _buildMessage(nonce) {
        const origin = window.location.origin;
        if (typeof this._config.messageBuilder === 'function') {
            return this._config.messageBuilder({
                appName: this._config.appName,
                origin,
                chainId: this._config.chainId,
                nonce,
            });
        }
        const domain = window.location.host;
        return [
            `${domain} wants you to sign in with your account:`,
            ``,
            `App: ${this._config.appName}`,
            `Nonce: ${nonce}`,
            `URI: ${origin}`,
            `Chain ID: ${this._config.chainId}`,
            `Version: 1`,
        ].join('\n');
    }
    /** Decode a message that may arrive as UTF-8 hex (0x...) or plain string */
    _decodeMessage(raw) {
        if (typeof raw !== 'string')
            return String(raw);
        if (raw.startsWith('0x')) {
            try {
                const hex = raw.slice(2);
                const bytes = new Uint8Array(hex.match(/.{1,2}/g).map((b) => parseInt(b, 16)));
                return new TextDecoder().decode(bytes);
            }
            catch {
                return raw;
            }
        }
        return raw;
    }
}
